ml-algorithms/src/main/java/org/opensearch/ml/engine/tools/SearchIndexTool.java

```diff
@@ -0,0 +1,195 @@
+    }
+
+    @Override
+    public String getType() {
+        return TYPE;
+    }
+
+    @Override
+    public String getVersion() {
+        return null;
+    }
+
+    @Override
+    public boolean validate(Map<String, String> parameters) {
+        if (parameters == null || !parameters.containsKey(INPUT_FIELD)) {
+            return false;
+        }
+        String input = parameters.get(INPUT_FIELD);
+        return input != null;
+    }
+
+    @Override
+    public <T> void run(Map<String, String> parameters, ActionListener<T> listener) {
+        try {
+            String input = parameters.get(INPUT_FIELD);
+            String index = "";
+            String query = "";
+            try {
+                // need this try catch block, because index, query field may not exist
+                JsonObject jsonObject = gson.fromJson(input, JsonObject.class);
+                index = jsonObject.get(INDEX_FIELD).getAsString();
+                query = jsonObject.get(QUERY_FIELD).toString();
+                query = "{\"query\": " + query + "}";
+            } catch (Exception e) {
+                // throw new IllegalArgumentException("wrong input");
+            }
+            if (org.apache.commons.lang3.StringUtils.isBlank(query)) {
+                throw new IllegalArgumentException("[" + QUERY_FIELD + "] is null or empty, can not process it.");
+            }
+
+            SearchSourceBuilder searchSourceBuilder = new SearchSourceBuilder();
+            XContentParser queryParser = XContentType.JSON
+                .xContent()
+                .createParser(xContentRegistry, LoggingDeprecationHandler.INSTANCE, query);
+            searchSourceBuilder.parseXContent(queryParser);
+            SearchRequest searchRequest = new SearchRequest().source(searchSourceBuilder).indices(index);
+
+            ActionListener actionListener = ActionListener.<SearchResponse>wrap(r -> {
+                SearchHit[] hits = r.getHits().getHits();
+
+                if (hits != null && hits.length > 0) {
+                    StringBuilder contextBuilder = new StringBuilder();
+                    for (int i = 0; i < hits.length; i++) {
+                        SearchHit hit = hits[i];
+                        String doc = AccessController.doPrivileged((PrivilegedExceptionAction<String>) () -> {
+                            Map<String, Object> docContent = new HashMap<>();
+                            docContent.put("_index", hit.getIndex());
+                            docContent.put("_id", hit.getId());
+                            docContent.put("_score", hit.getScore());
+                            docContent.put("_source", hit.getSourceAsMap());
+                            return gson.toJson(docContent);
+                        });
+                        contextBuilder.append(doc).append("\n");
+                    }
+                    listener.onResponse((T) gson.toJson(contextBuilder.toString()));
```

Comment:
Will the result be json? It seems like jsonl.

Reply:
json and jsonl, which one is preferred?

Reply:
json is standard and preferred.